sqlglot/dialects/duckdb.py

```diff
@@ -85,6 +85,21 @@ def _regexp_extract_sql(self: generator.Generator, expression: exp.RegexpExtract
     )
 
 
+def _json_format_sql(self: generator.Generator, expression: exp.JSONFormat) -> str:
+    this = expression.this
+    if not this.type:
+        from sqlglot.optimizer.annotate_types import annotate_types
+
+        annotate_types(this)
```

Comment:
What case triggered the need to annotate types here? Can't we always just cast to `TEXT`?

Reply:
this should only be done in the annotate types modules, let’s delete this

Reply:
I assume this whole section was copy-pasted from Hive?

https://github.com/tobymao/sqlglot/blob/4da37aa5caec407f66405b39f37d5dad057e66c2/sqlglot/dialects/hive.py#L87-L96

Reply:
IIRC the Hive approach used `annotate_types` because there was an issue with transpiling Trino's `JSON_FORMAT`, where we'd get back the invalid `TO_JSON(TO_JSON(..))` SQL in Spark (see https://github.com/tobymao/sqlglot/commit/aef9cfa9c for reference).

FWIW, it looks like we're still not transpiling Trino correctly to Spark, cc: @tobymao. For example:

```python
>>> import sqlglot
>>> sqlglot.transpile("SELECT JSON_FORMAT(JSON '[1, 2, 3]')", "trino", "spark")
["SELECT TO_JSON('[1, 2, 3]')"]
```

The resulting SQL here is not valid Spark:

```
spark-sql (default)> SELECT TO_JSON('[1, 2, 3]');
[DATATYPE_MISMATCH.INVALID_JSON_SCHEMA] Cannot resolve "to_json([1, 2, 3])" due to data type mismatch: Input schema "STRING" must be a struct, an array or a map.; line 1 pos 7;
```

I'm not sure if we can always transpile this function correctly without having a JSON parser, but for cases where we have nested (*) data structures represented as JSON strings, maybe we can alter the logic of `_json_format_sql` in Hive to output something like the following:

```python
>>> import sqlglot
>>> sqlglot.transpile("SELECT JSON_FORMAT(JSON '[1, 2, 3]')", "trino", "spark")
["SELECT TO_JSON(FROM_JSON('[1, 2, 3]', SCHEMA_OF_JSON('[1, 2, 3]')))"]
```

This is now valid Spark:

```
spark-sql (default)> SELECT TO_JSON(FROM_JSON('[1, 2, 3]', SCHEMA_OF_JSON('[1, 2, 3]')));
to_json(from_json([1, 2, 3]))
[1,2,3]
Time taken: 0.261 seconds, Fetched 1 row(s)
```

Similarly, but for a JSON string that encodes a map structure:

```
spark-sql (default)> SELECT TO_JSON(FROM_JSON('{"a": 1, "b": "c"}', SCHEMA_OF_JSON('{"a": 1, "b": "c"}')));
to_json(from_json({"a": 1, "b": "c"}))
{"a":1,"b":"c"}
Time taken: 4.434 seconds, Fetched 1 row(s)
```

Note that the outer `TO_JSON` call was added to cast it into a string again, which is what Trino's `JSON_FORMAT` [function](https://github.com/tobymao/sqlglot/assets/46752250/a250bf73-3c62-456b-8e2f-81062c4a231a) does.

```python
>>> from pyspark.sql import SQLContext
>>> from pyspark.context import SparkContext
>>> SQLContext(SparkContext()).sql("""SELECT TO_JSON(FROM_JSON('{"a": 1, "b": "c"}', SCHEMA_OF_JSON('{"a": 1, "b": "c"}')))""")
DataFrame[to_json(from_json({"a": 1, "b": "c"})): string]
```

I can try to implement this in another PR.

(*) I'm saying "nested" because e.g. for `'"a"'` this conversion fails:

```
spark-sql (default)> SELECT TO_JSON(FROM_JSON('"a"', SCHEMA_OF_JSON('"a"')));
[DATATYPE_MISMATCH.INVALID_JSON_SCHEMA] Cannot resolve "from_json("a")" due to data type mismatch: Input schema "STRING" must be a struct, an array or a map.; line 1 pos 15;
```

Reply:
sounds good